csrc/dynamic_transform.cpp

```diff
@@ -0,0 +1,477 @@
+        "Invalid DynamicTransformInfo. The associated Fusion is different from the given Fusion");
+    concretize();
+  }
+
+ private:
+  void concretize();
+
+  void concretizeReshape();
+
+  using OptOutMutator::mutate;
+
+  void mutate(TensorView* tv) final;
+
+  void mutate(TensorDomain* td) final;
+
+  //! Concretizes the root domain of a symbolic consumer tensor from
+  //! its producer domains. Returns true if any root ID is concretized.
+  bool propagateFromProducerToConsumer(TensorView* consumer);
+
+ private:
+  const DynamicTransformConcretizationInfo& info_;
+  std::unordered_map<IterDomain*, IterDomain*> update_map_;
+};
+
+void DynamicTransformConcretizer::concretize() {
+  // First, concretize all dynamic reshape ops
+  concretizeReshape();
+
+  // Second, propagate concretized domains
+  auto all_stmts = StmtSort::getStmts(info_.fusion(), false);
+  for (auto stmt : all_stmts) {
+    if (stmt->isA<Val>()) {
+      mutate(stmt);
+    }
+  }
+}
+
+void DynamicTransformConcretizer::concretizeReshape() {
+  // Concretize each reshape op.
+  for (const auto& kv : info_.getReshapeTransforms()) {
+    auto incomplete_out_tv = kv.first;
+    const auto view_analysis = kv.second;
+
+    auto inp_tv = ir_utils::producerTvsOf(incomplete_out_tv).at(0);
+
+    auto concrete_reshape_out_tv = reshape(inp_tv, view_analysis);
+
+    // Replace the old tensor with the new concretized tensor
+    for (auto use_of_old_tv : incomplete_out_tv->uses()) {
+      ir_utils::replaceValInExpr(
+          use_of_old_tv, incomplete_out_tv, concrete_reshape_out_tv);
+    }
+
+    if (incomplete_out_tv->isFusionOutput()) {
+      incomplete_out_tv->fusion()->replaceOutput(
+          incomplete_out_tv, concrete_reshape_out_tv);
+    }
+
+    incomplete_out_tv->fusion()->removeVal(incomplete_out_tv);
+  }
+}
+
+// Concretizes inherited symbolic domains. Note that when this is
+// called, it is assumed that all dynamic ops themselves are
+// concretized. Since symbolic IDs may be propagated down to
+// consumers, those domains need to be concretized accordingly.
+void DynamicTransformConcretizer::mutate(TensorView* tv) {
+  if (!tv->domain()->hasSymbolicAxis()) {
+    return;
+  }
+
+  // First, try to concretize the root domain as there may be symbolic
+  // axes inherited from the producers
+  auto propagated = propagateFromProducerToConsumer(tv);
+
+  // If no root domain is altered, nothing to do further
+  if (!propagated) {
+    return;
+  }
+
+  // Root IDs are altered. Need to propagate the changes to rfactor
+  // domain
+
+  // At this point, there should be no expr beyond rfactor root
+  TORCH_INTERNAL_ASSERT(
+      tv->domain()->domain() == tv->getMaybeRFactorDomain(),
+      "Invalid tensor: ",
+      tv->toString());
+
+  // If it has an rfactor root domain, the IterTypes of the rfactor
+  // IDs may need to be updated as well. Traverse the rfactor exprs
+  // and mutate the IterTypes of output IDs if symbolic.
+  if (tv->hasRFactor()) {
```

Comment:
@liqiangxl just an example of a traversal on transformations. You might not understand what's going on here, but this is a typical structure of such a traversal.